vendor/bat-native-ads/src/bat/ads/internal/processors/contextual/text_embedding/text_embedding_processor.cc

```diff
@@ -0,0 +1,122 @@
+#include "bat/ads/internal/tabs/tab_manager.h"
+#include "url/gurl.h"
+
+namespace ads {
+namespace processor {
+
+TextEmbedding::TextEmbedding(resource::TextEmbedding* resource)
+    : resource_(resource) {
+  DCHECK(resource_);
+
+  LocaleManager::GetInstance()->AddObserver(this);
+  ResourceManager::GetInstance()->AddObserver(this);
+  TabManager::GetInstance()->AddObserver(this);
+}
+
+TextEmbedding::~TextEmbedding() {
+  LocaleManager::GetInstance()->RemoveObserver(this);
+  ResourceManager::GetInstance()->RemoveObserver(this);
+  TabManager::GetInstance()->RemoveObserver(this);
+}
+
+bool TextEmbedding::IsEmbeddingEnabled() {
+  return targeting::features::IsTextEmbeddingEnabled();
+}
+
+void TextEmbedding::Process(const std::string& text) {
+  if (!resource_->IsInitialized()) {
+    BLOG(1, "Failed to process text embeddings as resource not initialized");
+    return;
+  }
+
+  const std::string sanitized_text = SanitizeHtml(text);
+  if (sanitized_text.length() == 0) {
+    BLOG(1, "No text available for embedding");
+    return;
+  }
+
+  ml::pipeline::EmbeddingProcessing* embedding_proc_pipeline = resource_->Get();
+  ml::pipeline::TextEmbeddingInfo text_embedding_info =
+      embedding_proc_pipeline->EmbedText(sanitized_text);
+  if (text_embedding_info.embedding.GetNonZeroElementCount() == 0) {
+    BLOG(1, "Failed to embed text");
+    return;
+  }
+
+  const std::string embedding_formatted =
+      text_embedding_info.embedding.GetVectorAsString();
+  BLOG(9, "Embedding: " << embedding_formatted);
+  LogTextEmbeddingHtmlEvent(
+      embedding_formatted, text_embedding_info.text_hashed,
+      [](const bool success) {
+        if (!success) {
+          BLOG(1, "Failed to log text embedding HTML event");
+          return;
+        }
+
+        PurgeStaleTextEmbeddingHtmlEvents([](const bool success) {
+          if (!success) {
+            BLOG(1, "Failed to purge stale text embedding HTML events");
+            return;
+          }
+        });
+      });
+}
+
+///////////////////////////////////////////////////////////////////////////////
+
+void TextEmbedding::OnLocaleDidChange(const std::string& /*locale*/) {
+  resource_->Load();
+}
+
+void TextEmbedding::OnResourceDidUpdate(const std::string& id) {
+  if (kLanguageComponentIds.find(id) != kLanguageComponentIds.end()) {
+    resource_->Load();
+  }
+}
+
+void TextEmbedding::OnHtmlContentDidChange(
+    const int32_t /*tab_id*/,
+    const std::vector<GURL>& redirect_chain,
+    const std::string& html) {
+  const GURL& url = redirect_chain.back();
+  DCHECK(!redirect_chain.empty());
+
+  if (!url.SchemeIsHTTPOrHTTPS()) {
+    BLOG(1, url.scheme() << " scheme is not supported for processing HTML");
+    return;
+  }
+
+  if (IsSearchEngine(url) && !IsSearchEngineResultsPage(url)) {
+    BLOG(1,
+         "Search engine landing pages are not supported for processing HTML");
+    return;
+  }
+
+  if (TextEmbedding::IsEmbeddingEnabled()) {
```

Comment:
Should this check not be moved to `targeting::BuildUserModel`?

Reply:
@tmancey I think this may be the case in the subsequent PR (@LorenzoMinto will be adding functionality to take the stored embeddings and transform them into some attribute of the user). As of now nothing is being written to a `UserModelInfo` object

Reply:
UserModelInfo is set from top segment matches.